docs/20_smart-contracts/40_guides/10_create_smart_contracts_guide.md

```diff
@@ -0,0 +1,1190 @@
+
+```cpp
+ACTION hello::addmsg(name nm, std::string message) {
+
+    require_auth( nm );
+
+    messages_table messages(get_self(), get_self().value);
+
+    messages.emplace(get_self(), [&](auto& row) {
+        row.user = nm;
+        row.message = message;
+        row.messagecks = eosio::sha256(message.data(), message.size());
+        row.time = current_time_point().time_since_epoch().count();
+    });
+}
+```
+
+Compile and deploy the smart contract, send the `addmsg` action with first parameter `ama` and sign it with `hello@active` key, and then observe how it fails with a standard error message:
+
+```cpp
+dune --send-action hello addmsg '[ama, "hi again"]' hello@active
+```
+
+```txt
+failed transaction: 8887cca7fababeb883aa6806c220eece3d5f3c618824b2378ac25a67c09a063a  <unknown> bytes  <unknown> us
+error 2023-03-01T15:53:35.033 cleos     main.cpp:700                  print_result         ] soft_except->to_detail_string(): 3090004 missing_auth_exception: Missing required authority
+missing authority of ama
+    {"account":"ama"}
+    nodeos  apply_context.cpp:256 require_authorization
+pending console output: 
+    {"console":""}
+    nodeos  apply_context.cpp:124 exec_one
+```
+
+### Use require_auth2()
+
+The `require_auth2()` enforces the execution only by the account that is set as the first parameter and only if the permission used to sign the transaction is the one specified as the second parameter. If the check fails it raises a standard error message which can not be customized.
+
+```cpp
+ACTION hello::addmsg(name nm, std::string message) {
+
+   require_auth2(nm.value, "active"_n.value);
+
+   messages_table messages(get_self(), get_self().value);
+
+   messages.emplace(get_self(), [&](auto& row) {
+      row.user = nm;
+      row.message = message;
+      row.messagecks = eosio::sha256(message.data(), message.size());
+      row.time = current_time_point().time_since_epoch().count();
+   });
+}
+```
+
+Compile and deploy the smart contract, send the `addmsg` action with first parameter `ama`, sign it with `ama@owner` private key, and then observe how it fails with a standard error message:
+
+```shell
+dune --send-action hello addmsg '[ama, "hi again"]' ama@owner
+```
+
+Even if the `ama@owner` private key was used to sign the above transaction, the execution fails because the required signature is the `ama@active`.
+
+```txt
+failed transaction: 7dcb10621e4102ec933cdbaf544f0204446cc96bc20b76242391b17286f1408e  <unknown> bytes  <unknown> us
+error 2023-03-01T16:00:02.853 cleos     main.cpp:700                  print_result         ] soft_except->to_detail_string(): 3090004 missing_auth_exception: Missing required authority
+missing authority of ama/active
+    {"account":"ama","permission":"active"}
+    nodeos  apply_context.cpp:275 require_authorization
+pending console output: 
+    {"console":""}
+    nodeos  apply_context.cpp:124 exec_one
+```
+
+## Assertions
```

Comment:
This should really be moved above authorizations since it uses `check` in the `has_auth` section.